src/extractor/guidance/coordinate_extractor.cpp

```diff
@@ -0,0 +1,834 @@
+        const auto &geometry = compressed_geometries.GetBucketReference(turn_edge);
+        result.reserve(geometry.size() + 2);
+
+        // the compressed edges contain node ids, we transfer them to coordinates accessing the
+        // node_coordinates array
+        const auto compressedGeometryToCoordinate = [this](
+            const CompressedEdgeContainer::OnewayCompressedEdge &compressed_edge) -> util::Coordinate {
+            return node_coordinates[compressed_edge.node_id];
+        };
+
+        // add the coordinates to the result in either normal or reversed order, based on
+        // traversed_in_reverse
+        if (traversed_in_reverse)
+        {
+            std::transform(geometry.rbegin(),
+                           geometry.rend(),
+                           std::back_inserter(result),
+                           compressedGeometryToCoordinate);
+            result.push_back(node_coordinates[intersection_node]);
+        }
+        else
+        {
+            result.push_back(node_coordinates[intersection_node]);
+            std::transform(geometry.begin(),
+                           geometry.end(),
+                           std::back_inserter(result),
+                           compressedGeometryToCoordinate);
+        }
+        return result;
+    }
+}
+
+double
+CoordinateExtractor::GetMaxDeviation(std::vector<util::Coordinate>::const_iterator range_begin,
+                                     const std::vector<util::Coordinate>::const_iterator &range_end,
+                                     const util::Coordinate &straight_begin,
+                                     const util::Coordinate &straight_end) const
+{
+    double deviation = 0;
+    for (; range_begin != range_end; range_begin = std::next(range_begin))
+    {
+        // find the projected coordinate
+        auto coord_between = util::coordinate_calculation::projectPointOnSegment(
+                                 straight_begin, straight_end, *range_begin)
+                                 .second;
+        // and calculate the distance between the intermediate coordinate and the coordinate
+        // on the osrm-way
+        deviation =
+            std::max(deviation,
+                     util::coordinate_calculation::haversineDistance(coord_between, *range_begin));
+    }
+    return deviation;
+};
+
+bool CoordinateExtractor::IsCurve(const std::vector<util::Coordinate> &coordinates,
+                                  const std::vector<double> &segment_distances,
+                                  const double segment_length,
+                                  const double considered_lane_width,
+                                  const util::NodeBasedEdgeData &edge_data) const
+{
+    BOOST_ASSERT(coordinates.size() > 2);
+
+    // by default, we treat roundabout as curves
+    if (edge_data.roundabout)
+        return true;
+
+    // TODO we might have to fix this to better compensate for errors due to repeated coordinates
```

Comment:
Circle fitting came indeed to mind during writing this. I am unsure how to differentiate correctly between slight curves, minor errors in putting in roads. Could be a candidate to consider in further improvements.